contracts/DsaWrapper.sol

```diff
@@ -1,62 +1,244 @@
+        bytes4 basicDeposit = bytes4(
+            keccak256("deposit(address,uint256,uint256,uint256)")
+        );
+
+        data[0] = abi.encodeWithSelector(
+            basicDeposit,
+            _tokenAddress,
+            _amt,
+            0,
+            0
+        );
+        IDSA(dsaAddress).cast(targets, data, address(this));
+    }
+
+    /**
+     * @dev withdraws ERC20 token from DSA account given by _id
+     * @param _id id of DSA account
+     * @param _amt amount of tokens to withdraw
+     * @param _tokenAddress contract address of ERC20 token
+     */
+    function withdrawErc20(
+        uint256 _id,
+        uint256 _amt,
+        address _tokenAddress
+    ) external onlyAuthority(_id, msg.sender) {
+        IERC20 token = IERC20(_tokenAddress);
+        
+        address dsaAddress = instapool.getAccountIdDetails(_id).account;
+
+        require(token.balanceOf(dsaAddress) >= _amt, "INSUFFICIENT TOKEN BALANCE");
+
+        string[] memory targets = new string[](1);
+        targets[0] = "BASIC-A";
+
+        bytes[] memory data = new bytes[](1);
+        bytes4 basicWithdraw = bytes4(
+            keccak256("withdraw(address,uint256,address,uint256,uint256)")
+        );
+
+        data[0] = abi.encodeWithSelector(
+            basicWithdraw,
+            _tokenAddress,
+            _amt,
+            msg.sender,
+            0,
+            0
+        );
+
+        IDSA(dsaAddress).cast(targets, data, address(this));
     }
 
-    // withdraw ether/erc20 from dsa
+    /**
+     * @dev adds the passed address to the list of authority accounts of the DSA
+     * @param _id id of the DSA account
+     * @param _authority address of account to be added as authority to DSA
+     */
+    function addAuthority(uint256 _id, address _authority)
+        external
+        onlyAuthority(_id, msg.sender)
+    {
+        require(_authority != address(0), "INVALID ADDRESS");
 
-    // add authority
+        address dsaAddress = instapool.getAccountIdDetails(_id).account;
+        string[] memory target = new string[](1);
+        target[0] = "AUTHORITY-A";
+
+        bytes[] memory data = new bytes[](1);
+        bytes4 addAuth = bytes4(keccak256("add(address)"));
 
-    // remove authority
+        data[0] = abi.encodeWithSelector(addAuth, _authority);
+        IDSA(dsaAddress).cast(target, data, address(this));
+    }
+
+    /**
+     * @dev removes the passed address from the list of authority accounts of the DSA
+     * @param _id id of the DSA account
+     * @param _authority address of account to be removed from authorities to DSA
+     */
+    function removeAuthority(uint256 _id, address _authority)
+        external
+        onlyAuthority(_id, msg.sender)
+    {
+        require(_authority != address(0), "INVALID ADDRESS");
 
-}
\ No newline at end of file
+        address dsaAddress = instapool.getAccountIdDetails(_id).account;
+
+        string[] memory target = new string[](1);
+        target[0] = "AUTHORITY-A";
+
```

Comment:
There should be one more check here that he can't remove all the authorities.
At least one authority should be there.

Reply:
Since the function only allows removing a single authority, I can just add a check to see if the address being removed is not the last authority for the DSA account. Is this okay, or should I allow removing more than one authority and place a check accordingly?

Reply:
Yes, check that the user shouldn't be removing the final eft authority.